Implement trace function

FikriMilano opened the pull request:
Fixes #40 

- `trace()` prints the current fhirpath context for debugging purposes.
- to be able to see the logs, `showStandardStreams = true` is needed.
- new API to acces the logs `engine.traces` (read only)

<img width="805" height="218" alt="Screen Shot 2025-12-23 at 15 48 22" src="https://github.com/user-attachments/assets/53525132-b385-47e9-8499-9c5aee37982c" />

FikriMilano committed:
Implement trace function

LZRS reviewed (commented):
The PR looks good. 

Other than just logging the results for the `trace` fhirpath function, could there be a mechanism that allows for getting back the results?

In [FhirPath Lab](https://fhirpath-lab.com/FhirPath), they have a requirement that whenever the `trace` function is used, the `Parameters` result returned from the fhirpath server should have a parameter with the label of the trace and its value
 
Example of a request sent from Fhirpath labs

```json
{
    "resourceType": "Parameters",
    "parameter": [
        {
            "name": "expression",
            "valueString": "trace('trc').given.join(' ')\n.combine(family).join(', ')"
        },
        {
            "name": "context",
            "valueString": "name"
        },
        {
            "name": "resource",
            "resource": {
                "resourceType": "Patient",
                "id": "example",
                "active": true,
                "name": [
                    {
                        "use": "official",
                        "family": "Chalmers",
                        "given": [
                            "Peter",
                            "James"
                        ]
                    },
                    {
                        "use": "usual",
                        "given": [
                            "Jim"
                        ]
                    },
                    {
                        "use": "maiden",
                        "family": "Windsor",
                        "given": [
                            "Peter",
                            "James"
                        ],
                        "period": {
                            "end": "2002"
                        }
                    }
                ]
            }
        },
        {
            "name": "terminologyserver",
            "valueString": "https://sqlonfhir-r4.azurewebsites.net/fhir"
        }
    ]
}
```

Expected Response

```json
{
    "resourceType": "Parameters",
    "id": "fhirpath",
    "parameter": [
        {
            "name": "parameters",
            "part": [
                {
                    "name": "evaluator",
                    "valueString": "fhirpath-py 2.1.0"
                },
                {
                    "name": "context",
                    "valueString": "name"
                },
                {
                    "name": "expression",
                    "valueString": "trace('trc').given.join(' ')\n.combine(family).join(', ')"
                },
                {
                    "name": "resource",
                    "resource": {
                        "resourceType": "Patient",
                        "id": "example",
                        "active": true,
                        "name": [
                            {
                                "use": "official",
                                "family": "Chalmers",
                                "given": [
                                    "Peter",
                                    "James"
                                ]
                            },
                            {
                                "use": "usual",
                                "given": [
                                    "Jim"
                                ]
                            },
                            {
                                "use": "maiden",
                                "family": "Windsor",
                                "given": [
                                    "Peter",
                                    "James"
                                ],
                                "period": {
                                    "end": "2002"
                                }
                            }
                        ]
                    }
                },
                {
                    "name": "terminologyServerUrl",
                    "valueString": "https://sqlonfhir-r4.azurewebsites.net/fhir"
                }
            ]
        },
        {
            "name": "result",
            "part": [
                {
                    "name": "str",
                    "valueString": "\"Peter James, Chalmers\""
                },
                {
                    "name": "trace",
                    "valueString": "trc",
                    "part": [
                        {
                            "extension": [
                                {
                                    "url": "http://fhir.forms-lab.com/StructureDefinition/resource-path",
                                    "valueString": "Patient.name[0]"
                                }
                            ],
                            "name": "HumanName",
                            "valueHumanName": {
                                "use": "official",
                                "family": "Chalmers",
                                "given": [
                                    "Peter",
                                    "James"
                                ]
                            }
                        }
                    ]
                }
            ],
            "valueString": "Patient.name[0]"
        },
        {
            "name": "result",
            "part": [
                {
                    "name": "str",
                    "valueString": "\"Jim\""
                },
                {
                    "name": "trace",
                    "valueString": "trc",
                    "part": [
                        {
                            "extension": [
                                {
                                    "url": "http://fhir.forms-lab.com/StructureDefinition/resource-path",
                                    "valueString": "Patient.name[1]"
                                }
                            ],
                            "name": "HumanName",
                            "valueHumanName": {
                                "use": "usual",
                                "given": [
                                    "Jim"
                                ]
                            }
                        }
                    ]
                }
            ],
            "valueString": "Patient.name[1]"
        },
        {
            "name": "result",
            "part": [
                {
                    "name": "str",
                    "valueString": "\"Peter James, Windsor\""
                },
                {
                    "name": "trace",
                    "valueString": "trc",
                    "part": [
                        {
                            "extension": [
                                {
                                    "url": "http://fhir.forms-lab.com/StructureDefinition/resource-path",
                                    "valueString": "Patient.name[2]"
                                }
                            ],
                            "name": "HumanName",
                            "valueHumanName": {
                                "use": "maiden",
                                "family": "Windsor",
                                "given": [
                                    "Peter",
                                    "James"
                                ],
                                "period": {
                                    "end": "2002"
                                }
                            }
                        }
                    ]
                }
            ],
            "valueString": "Patient.name[2]"
        }
    ]
}
```

At the moment, it's not possible to get back the evaluated values of the `trace` functions

FikriMilano committed:
Return actual logs when trace is used

FikriMilano committed:
Create new API that returns trace logs

FikriMilano committed:
Test returned trace logs

FikriMilano committed:
Merge branch 'main' of github.com:google/kotlin-fhirpath into 40-trace-function

FikriMilano committed:
Re-mplement trace logs return after merge conflicts

- Use engine.trace since FhirPathEngine is now a class, before it was just a file with functions.
- If the same trace key is found i.e. 2 trace with same label is used on 1 expression, results will be flattened.

FikriMilano committed:
Include path in the stdout

trace[families]: [Patient.name[0]: Chalmers, Patient.name[1]: Windsor]

LZRS commented on fhir-path/src/commonMain/kotlin/com/google/fhir/fhirpath/FhirPathEvaluator.kt:
```diff
@@ -68,6 +70,8 @@ internal class FhirPathEvaluator(
   private val totalStack = ArrayDeque<Collection<Any>>()
   private val variables = mutableMapOf<String, Any?>()
+  private val _traces = mutableMapOf<String, MutableList<TraceEntry>>()
+  val traces: Map<String, List<TraceEntry>> get() = _traces
```
Create and return a new readonly copy here instead of reference, such that any new changes to the underlying `_traces` doesn't affect it.

FikriMilano replied:
I'm changing this to just 1 variable with `private set`. Property will be exposed to the public but only this class can update it, which exactly what we need in this case.

https://kotlinlang.org/docs/properties.html#changing-visibility-or-adding-annotations

FikriMilano committed:
Use private set for the `traces` property

FikriMilano committed:
Merge branch 'main' of github.com:google/kotlin-fhirpath into 40-trace-function

LZRS commented on fhir-path/src/jvmTest/kotlin/dev/ohs/fhir/fhirpath/TraceTest.jvm.kt:
```diff
@@ -0,0 +1,123 @@
+ * limitations under the License.
+ */
+
+package com.google.fhir.fhirpath
+
+import com.google.fhir.model.r4.FhirR4Json
+import com.google.fhir.model.r4.Resource
```
Replace `com.google.fhir` with `dev.ohs.fhir`, causing build errors in `main`

FikriMilano replied:
closed by https://github.com/ohs-foundation/kotlin-fhirpath/issues/77
